Felix Shyann

In [1]:
import pandas as pd
from sklearn.model_selection import cross_val_score
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.ensemble import RandomForestClassifier

df = pd.read_csv('diabetes_012_health_indicators_BRFSS2015.csv')

df



,Diabetes_012,HighBP,HighChol,CholCheck,BMI,Smoker,Stroke,HeartDiseaseorAttack,PhysActivity,Fruits,...,AnyHealthcare,NoDocbcCost,GenHlth,MentHlth,PhysHlth,DiffWalk,Sex,Age,Education,Income
0,0.0,1.0,1.0,1.0,40.0,1.0,0.0,0.0,0.0,0.0,...,1.0,0.0,5.0,18.0,15.0,1.0,0.0,9.0,4.0,3.0
1,0.0,0.0,0.0,0.0,25.0,1.0,0.0,0.0,1.0,0.0,...,0.0,1.0,3.0,0.0,0.0,0.0,0.0,7.0,6.0,1.0
2,0.0,1.0,1.0,1.0,28.0,0.0,0.0,0.0,0.0,1.0,...,1.0,1.0,5.0,30.0,30.0,1.0,0.0,9.0,4.0,8.0
3,0.0,1.0,0.0,1.0,27.0,0.0,0.0,0.0,1.0,1.0,...,1.0,0.0,2.0,0.0,0.0,0.0,0.0,11.0,3.0,6.0
4,0.0,1.0,1.0,1.0,24.0,0.0,0.0,0.0,1.0,1.0,...,1.0,0.0,2.0,3.0,0.0,0.0,0.0,11.0,5.0,4.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
253675,0.0,1.0,1.0,1.0,45.0,0.0,0.0,0.0,0.0,1.0,...,1.0,0.0,3.0,0.0,5.0,0.0,1.0,5.0,6.0,7.0
253676,2.0,1.0,1.0,1.0,18.0,0.0,0.0,0.0,0.0,0.0,...,1.0,0.0,4.0,0.0,0.0,1.0,0.0,11.0,2.0,4.0
253677,0.0,0.0,0.0,1.0,28.0,0.0,0.0,0.0,1.0,1.0,...,1.0,0.0,1.0,0.0,0.0,0.0,0.0,2.0,5.0,2.0
253678,0.0,1.0,0.0,1.0,23.0,0.0,0.0,0.0,0.0,1.0,...,1.0,0.0,3.0,0.0,0.0,0.0,1.0,7.0,5.0,1.0


In [2]:
df = df.drop_duplicates(keep='first')
df.shape

(229781, 22)

------------------

**A- Define the prediction task:**

Outcome variable: diabetes_012: Whether or not the individual has diabetes or pre diabetes

2= diabetic

1= Pre diabetic

0= individual does not have diabetes



**B- Predictor variables: these may help predict the outcome because...:**

1. Age: Younger individuals may not populate the diabetic column as much as older individuals would.

2. Smoker: Smoking generally isn't good for health, and according to the FDA, it increases insulin resistance.
 
3. BMI: A higher BMI raises the risk for diabetes and also increases insulin resistance.

4. PhysActivity: Physical activity betters health, and those who aren't physically active are at risk of being diabetic.
 
5. HighBP: High blood pressure increases insulin resistance, damages blood vessels, and increases risk for organ damage.

6. HighChol: high cholesterol and diabetes normally occur together and it increase the risk for diabetes.

-----------------------------------------

**Data Preparation**

In [3]:
#Convert one categorical feature into numeric using one hot encoding:

In [4]:
categorical_columns = ['Age']

diabetes_encoded = pd.get_dummies(df, columns=categorical_columns, dtype=float, drop_first=True)

diabetes_encoded.head()

,Diabetes_012,HighBP,HighChol,CholCheck,BMI,Smoker,Stroke,HeartDiseaseorAttack,PhysActivity,Fruits,...,Age_4.0,Age_5.0,Age_6.0,Age_7.0,Age_8.0,Age_9.0,Age_10.0,Age_11.0,Age_12.0,Age_13.0
0,0.0,1.0,1.0,1.0,40.0,1.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0,0.0,0.0
1,0.0,0.0,0.0,0.0,25.0,1.0,0.0,0.0,1.0,0.0,...,0.0,0.0,0.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0
2,0.0,1.0,1.0,1.0,28.0,0.0,0.0,0.0,0.0,1.0,...,0.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0,0.0,0.0
3,0.0,1.0,0.0,1.0,27.0,0.0,0.0,0.0,1.0,1.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0
4,0.0,1.0,1.0,1.0,24.0,0.0,0.0,0.0,1.0,1.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0


In [5]:
#Check if there are missing values:
diabetes_encoded.isnull().sum()

Diabetes_012            0
HighBP                  0
HighChol                0
CholCheck               0
BMI                     0
Smoker                  0
Stroke                  0
HeartDiseaseorAttack    0
PhysActivity            0
Fruits                  0
Veggies                 0
HvyAlcoholConsump       0
AnyHealthcare           0
NoDocbcCost             0
GenHlth                 0
MentHlth                0
PhysHlth                0
DiffWalk                0
Sex                     0
Education               0
Income                  0
Age_2.0                 0
Age_3.0                 0
Age_4.0                 0
Age_5.0                 0
Age_6.0                 0
Age_7.0                 0
Age_8.0                 0
Age_9.0                 0
Age_10.0                0
Age_11.0                0
Age_12.0                0
Age_13.0                0
dtype: int64

In [6]:
#No missing values found above

In [7]:
#Split data into training and testing sets
age_dummy_cols = [col for col in diabetes_encoded.columns if col.startswith('Age_')]
predictors = ['Smoker', 'BMI', 'PhysActivity', 'HighBP', 'HighChol'] + age_dummy_cols

X_diabetes = diabetes_encoded[predictors]
y_diabetes = diabetes_encoded['Diabetes_012']

In [8]:
X_diabetes

,Smoker,BMI,PhysActivity,HighBP,HighChol,Age_2.0,Age_3.0,Age_4.0,Age_5.0,Age_6.0,Age_7.0,Age_8.0,Age_9.0,Age_10.0,Age_11.0,Age_12.0,Age_13.0
0,1.0,40.0,0.0,1.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0,0.0,0.0
1,1.0,25.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0
2,0.0,28.0,0.0,1.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0,0.0,0.0
3,0.0,27.0,1.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0
4,0.0,24.0,1.0,1.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
253675,0.0,45.0,0.0,1.0,1.0,0.0,0.0,0.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
253676,0.0,18.0,0.0,1.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0
253677,0.0,28.0,1.0,0.0,0.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
253678,0.0,23.0,0.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0


In [9]:
from sklearn.model_selection import train_test_split

X_train_diabetes, X_test_diabetes, y_train_diabetes, y_test_diabetes = train_test_split(X_diabetes, y_diabetes, test_size=0.2, random_state=42)

In [10]:
#Balance the training data
y_train_diabetes.value_counts()

Diabetes_012
0.0    151939
2.0     28162
1.0      3723
Name: count, dtype: int64

The data set is imbalanced because the number of individuals without diabetes (0) is much higher than the number of individuals with diabetes (2) and prediabetes (1). The model may favor the largest group and have trouble predicting the smaller groups, so the training data should be balanced.

In [11]:
#balance
from imblearn.over_sampling import RandomOverSampler

randOverSample = RandomOverSampler(random_state=42)
X_train_diabetes_balanced, y_train_diabetes_balanced = randOverSample.fit_resample(X_train_diabetes, y_train_diabetes)

---------

**Model Selection with 5-fold Cross Validation**

**Initialize Logistic Regression and train a logistic regression model using 5-fold cross-validation:**

In [12]:
logreg = LogisticRegression(max_iter=2000)

accuracy_scores_logreg = cross_val_score(logreg, X_train_diabetes_balanced, y_train_diabetes_balanced, cv=5, scoring='accuracy')
print("Logistic Regression - Accuracy Scores across folds:", accuracy_scores_logreg)
print("Logistic Regression - Mean Accuracy:", accuracy_scores_logreg.mean())
print(" ")

precision_scores_logreg = cross_val_score(logreg, X_train_diabetes_balanced, y_train_diabetes_balanced, cv=5, scoring='precision_weighted')
print("Logistic Regression - Precision Scores across folds:", precision_scores_logreg)
print("Logistic Regression - Mean Precision:", precision_scores_logreg.mean())
print(" ")

recall_scores_logreg = cross_val_score(logreg, X_train_diabetes_balanced, y_train_diabetes_balanced, cv=5, scoring='recall_weighted')
print("Logistic Regression - Recall Scores across folds:", recall_scores_logreg)
print("Logistic Regression - Mean Recall:", recall_scores_logreg.mean())
print(" ")

f1_scores_logreg = cross_val_score(logreg, X_train_diabetes_balanced, y_train_diabetes_balanced, cv=5, scoring='f1_weighted')
print("Logistic Regression - F1 Scores across folds:", f1_scores_logreg)
print("Logistic Regression - Mean F1 Score:", f1_scores_logreg.mean())

Logistic Regression - Accuracy Scores across folds: [0.48393006 0.48682594 0.48708358 0.48492261 0.48564659]
Logistic Regression - Mean Accuracy: 0.48568175519454126
 
Logistic Regression - Precision Scores across folds: [0.47159146 0.47525001 0.47477543 0.47313142 0.47480513]
Logistic Regression - Mean Precision: 0.47391068928053415
 
Logistic Regression - Recall Scores across folds: [0.48393006 0.48682594 0.48708358 0.48492261 0.48564659]
Logistic Regression - Mean Recall: 0.48568175519454126
 
Logistic Regression - F1 Scores across folds: [0.46183144 0.46474717 0.46540817 0.4630691  0.46418966]
Logistic Regression - Mean F1 Score: 0.46384910820145964


-----

 **Initialize Decision tree and train it using 5-fold cross-validation:**

In [13]:
from sklearn.tree import DecisionTreeClassifier

tree = DecisionTreeClassifier(max_depth=10)

accuracy_scores_tree = cross_val_score(tree, X_train_diabetes_balanced, y_train_diabetes_balanced, cv=5, scoring='accuracy', n_jobs=-1)
print("Decision Tree - Accuracy Scores across folds:", accuracy_scores_tree)
print("Decision Tree - Mean Accuracy:", accuracy_scores_tree.mean())
print(" ")

precision_scores_tree = cross_val_score(tree, X_train_diabetes_balanced, y_train_diabetes_balanced, cv=5, scoring='precision_weighted', n_jobs=-1)
print("Decision Tree - Precision Scores across folds:", precision_scores_tree)
print("Decision Tree - Mean Precision:", precision_scores_tree.mean())
print(" ")

recall_scores_tree = cross_val_score(tree, X_train_diabetes_balanced, y_train_diabetes_balanced, cv=5, scoring='recall_weighted', n_jobs=-1)
print("Decision Tree - Recall Scores across folds:", recall_scores_tree)
print("Decision Tree - Mean Recall:", recall_scores_tree.mean())
print(" ")

f1_scores_tree = cross_val_score(tree, X_train_diabetes_balanced, y_train_diabetes_balanced, cv=5, scoring='f1_weighted', n_jobs=-1)
print("Decision Tree - F1 Scores across folds:", f1_scores_tree)
print("Decision Tree - Mean F1 Score:", f1_scores_tree.mean())

Decision Tree - Accuracy Scores across folds: [0.51334957 0.51343732 0.51438632 0.51155622 0.51328938]
Decision Tree - Mean Accuracy: 0.5132037629588988
 
Decision Tree - Precision Scores across folds: [0.51436867 0.51628014 0.51360667 0.51337244 0.51474498]
Decision Tree - Mean Precision: 0.5144745797284347
 
Decision Tree - Recall Scores across folds: [0.51334957 0.51343732 0.51438632 0.51155622 0.51328938]
Decision Tree - Mean Recall: 0.5132037629588988
 
Decision Tree - F1 Scores across folds: [0.51339253 0.51454108 0.51378243 0.51129929 0.51358833]
Decision Tree - Mean F1 Score: 0.513320733625461


----

**Initialize random forest and train it using 5-fold cross-validation**

In [14]:
rf = RandomForestClassifier(n_estimators=100)

accuracy_scores_rf = cross_val_score(rf, X_train_diabetes_balanced, y_train_diabetes_balanced, cv=5, scoring='accuracy')
print("Random Forest - Accuracy Scores across folds:", accuracy_scores_rf)
print("Random Forest - Mean Accuracy:", accuracy_scores_rf.mean())
print(" ")

precision_scores_rf = cross_val_score(rf, X_train_diabetes_balanced, y_train_diabetes_balanced, cv=5, scoring='precision_weighted')
print("Random Forest - Precision Scores across folds:", precision_scores_rf)
print("Random Forest - Mean Precision:", precision_scores_rf.mean())
print(" ")

recall_scores_rf = cross_val_score(rf, X_train_diabetes_balanced, y_train_diabetes_balanced, cv=5, scoring='recall_weighted')
print("Random Forest - Recall Scores across folds:", recall_scores_rf)
print("Random Forest - Mean Recall:", recall_scores_rf.mean())
print(" ")

f1_scores_rf = cross_val_score(rf, X_train_diabetes_balanced, y_train_diabetes_balanced, cv=5, scoring='f1_weighted')
print("Random Forest - F1 Scores across folds:", f1_scores_rf)
print("Random Forest - Mean F1 Score:", f1_scores_rf.mean())

Random Forest - Accuracy Scores across folds: [0.58957483 0.58881795 0.58959227 0.58878054 0.58774942]
Random Forest - Mean Accuracy: 0.5889030014528117
 
Random Forest - Precision Scores across folds: [0.59598406 0.59426972 0.59574326 0.59493813 0.59416412]
Random Forest - Mean Precision: 0.595019857580937
 
Random Forest - Recall Scores across folds: [0.58931157 0.5893006  0.58960324 0.5891864  0.58759584]
Random Forest - Mean Recall: 0.5889995313621283
 
Random Forest - F1 Scores across folds: [0.58899693 0.58850142 0.58913874 0.58851482 0.58638323]
Random Forest - Mean F1 Score: 0.5883070259056769


------

**Final Model:**

**Retrain the best model on the whole training set and test it. Report accuracy, F1, Recall, and precision.**

In [15]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

rf.fit(X_train_diabetes_balanced, y_train_diabetes_balanced)
y_pred_diabetes = rf.predict(X_test_diabetes)

test_accuracy = accuracy_score(y_test_diabetes, y_pred_diabetes)
test_precision = precision_score(y_test_diabetes, y_pred_diabetes, average='weighted')
test_recall = recall_score(y_test_diabetes, y_pred_diabetes, average='weighted')
test_f1 = f1_score(y_test_diabetes, y_pred_diabetes, average='weighted')

print("Test Accuracy (Best Model):" + str(test_accuracy))
print("Test Precision (Best Model):" + str(test_precision))
print("Test Recall (Best Model):" + str(test_recall))
print("Test F1 Score (Best Model):" + str(test_f1))

Test Accuracy (Best Model):0.5375242074112757
Test Precision (Best Model):0.8117920063479469
Test Recall (Best Model):0.5375242074112757
Test F1 Score (Best Model):0.6306154365571086


Interpretation: The final Random Forest model performed well, achieving a test accuracy of 0.540, a precision of 0.812(a noticeable jump from 0.595, the initializing training model), a recall of 0.540, and an F1 score of 0.630. All RF performance scores are slightly higher than those of the other two models implemented above. Also, an accuracy of 0.54 means the model gets the correct category right about half the time, which is better than random guessing. Finally, I chose Random Forest as the best-performing model because it achieved the highest mean accuracy, precision, recall, and F1 across all 5 cross-validation folds, compared to Logistic Regression and Decision Tree.

---------

**Logistic Regression with L1 Regularization**

In [16]:
# Logistic Regression with Lasso (L1) regularization
# Used to make the table easy to read: https://docs.python.org/3/library/functions.html#zip
from sklearn.metrics import classification_report

lasso_model = LogisticRegression(penalty='l1', solver='liblinear', C=0.01, max_iter=1000)
lasso_model.fit(X_train_diabetes_balanced, y_train_diabetes_balanced)
print(lasso_model.coef_)
print(list(zip(X_diabetes.columns, lasso_model.coef_[0])))

lasso_pred = lasso_model.predict(X_test_diabetes)
print(" ")
print("Lasso (L1) Regularization Test Set Evaluation:")
print(classification_report(y_test_diabetes, lasso_pred))

[[-0.05200669 -0.07253441  0.23705968 -0.68607747 -0.65259522  0.43335063
   0.44873725  0.          0.         -0.3417621  -0.43526264 -0.54133252
  -0.74673733 -0.89004381 -1.01053687 -1.04226389 -1.00622616]
 [ 0.          0.01279969  0.00913268  0.0261343   0.30085719 -0.25281861
  -0.28281983  0.03131432 -0.03502949  0.24356098  0.1515681   0.16049006
   0.21704354  0.16984847  0.29264586  0.39383504  0.39574708]
 [ 0.06071345  0.04453027 -0.21437352  0.71478057  0.33456074 -0.71042023
  -0.5591265  -0.24083819 -0.02557208  0.          0.2188273   0.32690401
   0.4541816   0.62277659  0.59798757  0.52276598  0.50723435]]
[('Smoker', -0.05200669382331046), ('BMI', -0.07253441124206826), ('PhysActivity', 0.23705967593416896), ('HighBP', -0.686077471704519), ('HighChol', -0.6525952227458875), ('Age_2.0', 0.43335063100248344), ('Age_3.0', 0.4487372483643154), ('Age_4.0', 0.0), ('Age_5.0', 0.0), ('Age_6.0', -0.34176210193776324), ('Age_7.0', -0.4352626409084403), ('Age_8.0', -0.5413325

Interpretation: Compared to the best-performing model from 5-fold cross-validation, L1 performs better across all metrics. Random Forest, test set scores: accuracy 0.540, precision 0.812, recall 0.540, F1 0.630, where L1 scores an accuracy of 0.64, a weighted precision of 0.82, a weighted recall of 0.64, and a weighted F1 of 0.70. The variables that zeroed out were Age_4.0 and Age_5.0 (not diabetic), Smoker (pre-diabetic), and Age_6.0 (diabetic), and their removal makes sense to me because that information is likely the same or close to BMI, HighBP, and HighChol, since the rest of the variables already cover what those groups tell us.

------

**Logistic regression with L2 Regularization**

In [17]:
# Logistic Regression with Ridge (L2) regularization
ridge_model = LogisticRegression(penalty='l2', solver='liblinear')
ridge_model.fit(X_train_diabetes_balanced, y_train_diabetes_balanced)


print(ridge_model.coef_)
print(list(zip(X_diabetes.columns, ridge_model.coef_[0])))
      
ridge_pred = ridge_model.predict(X_test_diabetes)
print (" ")
print(classification_report(y_test_diabetes, ridge_pred))

[[-0.05095371 -0.07366075  0.23079603 -0.67532109 -0.64588946 -0.39797114
  -0.40916186 -0.92094328 -0.91917328 -1.26461161 -1.35604167 -1.46109761
  -1.66632214 -1.81060988 -1.93437967 -1.97154338 -1.93571773]
 [-0.00781461  0.01332717  0.02356188  0.01603321  0.29558549  0.48839684
   0.504397    0.89273916  0.77293796  1.09570294  0.99904999  1.00761171
   1.06325153  1.01578851  1.141752    1.24668715  1.24824153]
 [ 0.06240173  0.04555245 -0.20887257  0.70390645  0.32872572  0.18862661
   0.41524486  0.77664139  1.00735259  1.08028425  1.29724235  1.40321224
   1.52966362  1.69847699  1.67652057  1.60583847  1.59242459]]
[('Smoker', -0.050953707252025866), ('BMI', -0.07366074786900212), ('PhysActivity', 0.23079602875633887), ('HighBP', -0.6753210896944224), ('HighChol', -0.6458894557624612), ('Age_2.0', -0.3979711402661256), ('Age_3.0', -0.4091618631102701), ('Age_4.0', -0.920943276485957), ('Age_5.0', -0.9191732803485174), ('Age_6.0', -1.2646116090135333), ('Age_7.0', -1.35604166

Interpretation: L2 performed almost the same as L1, scoring an accuracy of 0.63, a precision of 0.82, a recall of 0.63, and an F1 score of 0.69. No coefficients were zeroed out here, as expected, since L2 only shrinks coefficients rather than eliminating/zeroing them. Being that L1 and L2 scored nearly the same, even after L1 dropped Age_4.0, Age_5.0, Smoker, and Age_6.0, makes me think that those variables didn't add much value.

----

**Logistic Regression with Elastic Net Regularization**

In [18]:
# Logistic Regression with Elastic Net regularization (Combination of L1 and L2)
elastic_net_model = LogisticRegression(penalty='elasticnet', C=0.05, solver='saga', l1_ratio=0.5)
elastic_net_model.fit(X_train_diabetes_balanced, y_train_diabetes_balanced)

print(elastic_net_model.coef_)
print(list(zip(X_diabetes.columns, elastic_net_model.coef_[1])))

elastic_net_pred = elastic_net_model.predict(X_test_diabetes)
print(" ")
print(classification_report(y_test_diabetes, elastic_net_pred))

[[-3.15628982e-02 -6.40362166e-02  1.61978377e-01 -4.60705477e-01
  -4.96913673e-01 -7.29575558e-04 -1.79188925e-02 -3.53457742e-01
  -4.07556781e-01 -6.35549786e-01 -7.23554817e-01 -7.96466467e-01
  -9.37171199e-01 -1.03254098e+00 -1.11587017e+00 -1.13896514e+00
  -1.11596632e+00]
 [ 0.00000000e+00  1.13877805e-07  0.00000000e+00  0.00000000e+00
   1.30717742e-01  1.84090783e-01  1.21527891e-01  2.86701966e-01
   1.29646864e-01  3.05622125e-01  2.20488654e-01  2.21161494e-01
   2.62799031e-01  2.44103095e-01  3.28650779e-01  3.92621706e-01
   3.84991647e-01]
 [ 4.88724614e-02  2.10918003e-02 -1.47285600e-01  4.80066717e-01
   1.78997125e-01 -3.11336642e-01 -1.57919503e-01  4.01081505e-02
   1.60159335e-01  1.94083228e-01  3.26659333e-01  3.96591212e-01
   4.87182794e-01  6.03731771e-01  6.00024849e-01  5.59140684e-01
   5.43779307e-01]]
[('Smoker', 0.0), ('BMI', 1.1387780503244406e-07), ('PhysActivity', 0.0), ('HighBP', 0.0), ('HighChol', 0.13071774204521944), ('Age_2.0', 0.1840907829

/opt/anaconda3/lib/python3.12/site-packages/sklearn/linear_model/_sag.py:349: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(


Interpretation: Elastic Net zeroed out Smoker, PhysActivity, and HighBP in the "pre-diabetic" row, indicating that those three variables add little value for identifying pre-diabetic cases, though they're still useful for the other two classes. When it comes to the performance, it scores a precision of 0.82, a recall of 0.62, an F1 of 0.68, and an accuracy of 0.62, which are all slightly below L1 (accuracy of 0.64, a weighted precision of 0.82, a weighted recall of 0.64, and a weighted F1 of 0.70,) and L2(accuracy of 0.63, a precision of 0.82, a recall of 0.63, and an F1 score of 0.69). Compared to the best model (random forest: accuracy 0.540, precision 0.812, recall 0.540, F1 0.630), Elastic Net scores higher across all categories. 

-----